snapcraft/pluginhandler.py

```diff
@@ -109,21 +110,41 @@ def makedirs(self):
     def notify_stage(self, stage, hint=''):
         logger.info('%s %s %s', stage, self.name, hint)
 
-    def is_dirty(self, stage):
-        try:
-            with open(self.statefile, 'r') as f:
-                lastStep = f.read()
-                return (common.COMMAND_ORDER.index(stage) >
-                        common.COMMAND_ORDER.index(lastStep))
-        except Exception:
-            return True
+    def last_step(self):
+        for step in reversed(common.COMMAND_ORDER):
+            if os.path.exists(self._step_state_file(step)):
+                return step
+
+        return None
+
+    def is_dirty(self, step):
+        last_step = self.last_step()
+        if last_step:
+            return (common.COMMAND_ORDER.index(step) >
```

Comment:
Yeah that would work, but doing so almost feels a little too clever to me. It'd be difficult to follow.